crates/polars-plan/src/plans/conversion/join.rs

```diff
@@ -145,6 +191,46 @@ pub fn resolve_join(
         InvalidOperation: "all join key expressions must be elementwise."
     );
 
+    // These are Arc<Schema>, into_owned is free.
+    let schema_left = schema_left.into_owned();
+    let schema_right = schema_right.into_owned();
+
+    let key_cols_coalesced =
+        options.args.should_coalesce() && matches!(&options.args.how, JoinType::Full);
+
+    if key_cols_coalesced {
```

Comment:
I've made it so that we maintain the input column types in the output for all cases except for full-join with coalesce=True